### PR TITLE
feat(python_samples): Added functionality & updated py sample template

### PR DESCRIPTION
**Pull request by runargs**
Updated template to be used for Python sample libraries, as well as adding tests and functionality to generate the template which will be utilized in the future but do not currently affect documentation generation.

### CODE REVIEW

File: synthtool/gcp/templates/python_samples/README.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
+[//]: # "This README.md file is auto-generated, all changes to this file will be lost."
+[//]: # "To regenerate it, use `python -m synthtool`."
+Java idiomatic client for [{{metadata['repo']['name_pretty']}}]
```
Comment (kolea2) [inline synthtool/gcp/templates/python_samples/README.md:3]: delete

File: synthtool/gcp/templates/python_samples/README.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+-------------------------------------------------------------------------------
+
+### Prerequisites
+See the [Authentication Getting Started Guide][authentication]. Contribution guidelines, the language style guide, and license information can be found in the base parent directiory.{% if metadata['repo']['requires_billing'] %}You will need to [enable billing][enable_billing] to use Google {{ metadata['repo']['name_pretty'] }}.{% endif %}
+{% if metadata['repo']['custom_content'] is defined %}{{ metadata['repo']['custom_content']}}{% endif %}
+If this is your first time working with GCP products, you will need to set up a project. *For more information on basic setup, view the information in the parent directory's README.*
+
+### Install Dependencies
+
+1. Check to make sure you have `pip` installed and have created a `virtualenv`.
```
Comment (kolea2) [inline synthtool/gcp/templates/python_samples/README.md:18]: check the numbering here?

File: synthtool/gcp/templates/python_samples/README.rst [hunk 3]
```diff
@@ -1,75 +0,0 @@
-{# The following line is a lie. BUT! Once jinja2 is done with it, it will
```
Comment (kolea2) [inline synthtool/gcp/templates/python_samples/README.rst:1]: Let's not delete these yet. Some products are still using this template.

Comment (kolea2) [reply synthtool/gcp/templates/python_samples/README.rst:1]: Same goes for other deleted files

File: synthtool/gcp/templates/python_samples/README.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+1. Install the dependencies needed to run the samples.
+
+```bash
+    $ pip install -r requirements.txt
+```
+
+## Samples
+-------------------------------------------------------------------------------
+{% if metadata['repo']['samples']|length %}
+Samples, quickstarts, and other documentation available for this product is available at <a href="{{ metadata['repo']['product_documentation'] }}">the product documentation.</a>.
```
Comment (kolea2) [inline synthtool/gcp/templates/python_samples/README.md:31]: is it possible to make the links similar to the ones listed at the bottom?

File: tests/test_sample_templates.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
+# Copyright 2018 Google LLC
```
Comment (kolea2) [inline tests/test_sample_templates.py:1]: 2020

**Review by kolea2 — CHANGES_REQUESTED** [review]

File: synthtool/gcp/templates/python_samples/README.md (hunk 2 above)
Comment (runargs) [reply synthtool/gcp/templates/python_samples/README.md:18]: Just a markdown syntax thing, all numbers get formatted to the correct order (shows on rich dif)

File: synthtool/gcp/templates/python_samples/README.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+-------------------------------------------------------------------------------
+
+### Prerequisites
+See the [Authentication Getting Started Guide][authentication]. Contribution guidelines, the language style guide, and license information can be found in the base parent directiory.{% if metadata['repo']['requires_billing'] %}You will need to [enable billing][enable_billing] to use Google {{ metadata['repo']['name_pretty'] }}.{% endif %}
+{% if metadata['repo']['custom_content'] is defined %}{{ metadata['repo']['custom_content']}}{% endif %}
+If this is your first time working with GCP products, you will need to set up a project. *For more information on basic setup, view the information in the parent directory's README.*
+
+### Install Dependencies
+
+1. Check to make sure you have `pip` installed and have created a `virtualenv`.
```
Comment (billyjacobson) [inline synthtool/gcp/templates/python_samples/README.md:18]: Link to instructions for making virtualenv

File: synthtool/gcp/templates/python_samples/README.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+1. Clone the sample folder and navigate to the sample directory you want to use.
```
Comment (billyjacobson) [inline synthtool/gcp/templates/python_samples/README.md:20]: s/Clone the sample folder/Clone the branch/

Also should people be cloning the branch or making a fork of it?

File: synthtool/gcp/templates/python_samples/README.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+[//]: # "To regenerate it, use `python -m synthtool`."
+
+## Python Sample Folder for {{ metadata['repo']['name_pretty'] }} 
+
+This directory contains samples for {{ metadata['repo']['name_pretty'] }}.
+
+## Getting Started
+-------------------------------------------------------------------------------
+
+### Prerequisites
```
Comment (billyjacobson) [inline synthtool/gcp/templates/python_samples/README.md:11]: What are these prerequisites for? Running the code? Contributing?

File: synthtool/gcp/templates/python_samples/README.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+See the [Authentication Getting Started Guide][authentication]. Contribution guidelines, the language style guide, and license information can be found in the base parent directiory.{% if metadata['repo']['requires_billing'] %}You will need to [enable billing][enable_billing] to use Google {{ metadata['repo']['name_pretty'] }}.{% endif %}
```
Comment (billyjacobson) [inline synthtool/gcp/templates/python_samples/README.md:12]: "Contribution guidelines, the language style guide, and license information can be found in the base parent directiory."

Perhaps a section for contributors would be good, then the main goal of the readme can focus towards consumers of the code, but still provide information for those who want to contribute

File: synthtool/gcp/templates/python_samples/README.md (hunk 6 above)
Comment (billyjacobson) [inline synthtool/gcp/templates/python_samples/README.md:18]: I'm not sure who this step is for. 

If you're a customer and use python you most likely have pip installed or know you need it if you see the pip install X command, and you probably already have whatever Python environment setup you'd need which may or may not be virtualenv.

If you're a contributor who isn't as familiar with python, this doesn't really provide enough information to help figure out what to do and should probably just be in the contributing docs.

File: synthtool/gcp/templates/python_samples/README.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+{% if metadata['repo']['custom_content'] is defined %}{{ metadata['repo']['custom_content']}}{% endif %}
+If this is your first time working with GCP products, you will need to set up a project. *For more information on basic setup, view the information in the parent directory's README.*
```
Comment (billyjacobson) [inline synthtool/gcp/templates/python_samples/README.md:14]: Is it the parent directory (e.g. python-docs-samples/bigtable/README) or the top-level directory (e.g. python-docs-samples/README)? Maybe just turn "set up a project." into a link to the guide

File: synthtool/gcp/templates/python_samples/README.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+### {{ metadata['repo']['samples'][sample]['name']}}
+-------------------------------------------------------------------------------
+
+<a href="https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/{{ metadata['repo']['repo'] }}&page=editor&open_in_editor={{ metadata['repo']['samples'][sample]['file'] }}">
+         <img alt="Open in Cloud Shell" src="http://gstatic.com/cloudssh/images/open-btn.png">
+</a>
+
+{{ metadata['repo']['samples'][sample]['description']}}
+
+To run this sample:
```
Comment (billyjacobson) [inline synthtool/gcp/templates/python_samples/README.md:43]: I know some python samples are runnable, but I'm under the impression that moving forward, the samples aren't supposed to be runnable. Not sure though

**Review by billyjacobson — CHANGES_REQUESTED** [review]

File: synthtool/gcp/common.py [hunk 12]
```diff
@@ -59,6 +59,12 @@ def _generic_library(self, directory: str, **kwargs) -> Path:
 
         return result
 
+    def py_samples(self, **kwargs) -> Path:
+        # kwargs["metadata"] is required to load values from .repo-metadata.json
+        if "metadata" not in kwargs:
+            kwargs["metadata"] = {}
```
Comment (SurferJeffAtGoogle) [inline synthtool/gcp/common.py:65]: Why is it necessary to set kwargs["metadata"] like this?

File: tests/test_sample_templates.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
+# Copyright 2020 Google LLC
```
Comment (SurferJeffAtGoogle) [inline tests/test_sample_templates.py:1]: Thank you for adding tests!

**Review by SurferJeffAtGoogle — CHANGES_REQUESTED** [review]

File: synthtool/gcp/templates/python_samples/README.md (hunk 7 above)
Comment (runargs) [reply synthtool/gcp/templates/python_samples/README.md:20]: Being pretty new to the Github flow I'm not exactly how to word this. What do you suggest?

File: synthtool/gcp/common.py (hunk 12 above)
Comment (runargs) [reply synthtool/gcp/common.py:65]: I'm not completely sure, I leveraged existing code from the rest of common.py. Perhaps [_generic_library()](https://github.com/googleapis/synthtool/blob/b962b480c5cf5146752ac50d0084fc53dec734af/synthtool/gcp/common.py#L47) isn't built to handle nonexistent metadata as is.

File: synthtool/gcp/templates/python_samples/README.md (hunk 7 above)
Comment (billyjacobson) [reply synthtool/gcp/templates/python_samples/README.md:20]: Cloning the branch is fine. 

You can say "1. Clone the branch and navigate to the sample directory you want to use."

File: synthtool/gcp/templates/python_samples/README.md (hunk 11 above)
Comment (runargs) [reply synthtool/gcp/templates/python_samples/README.md:21]: Added in ability to toggle this section if a sample is 'runnable'

File: synthtool/gcp/templates/python_samples/README.md (hunk 10 above)
Comment (runargs) [reply synthtool/gcp/templates/python_samples/README.md]: Turned "set up a project." into a link to the cloud.google.com SDK introductory docs

File: synthtool/gcp/templates/python_samples/README.md (hunk 6 above)
Comment (runargs) [reply synthtool/gcp/templates/python_samples/README.md]: This was a really good point, re: our chats I reassessed who the audience is for this README (mainly customers looking for a usage reference) and overhauled the organization of it overall to fit.

File: synthtool/gcp/templates/python_samples/README.md (hunk 9 above)
Comment (runargs) [reply synthtool/gcp/templates/python_samples/README.md]: Added and separated the contrib section :)

**Comment by runargs** [comment]
I created a demo of README generation available [here](https://github.com/runargs/python-bigtable/blob/synth-generation-DEMO/README.md). It's worth noting the only relevant file changes are to [README.md](https://github.com/runargs/python-bigtable/blob/synth-generation-DEMO/README.md) and [.repo.metadata.json](https://github.com/runargs/python-bigtable/blob/synth-generation-DEMO/.repo-metadata.json).

Future generation code changes will allow for this README to be generated specifically in the samples folder as opposed to where it is now, and some vestigial files (.rst files) were generated. These vestigial files are currently being generated because of READMEgen script, but will be decommissioned when this generation code is used.

File: synthtool/gcp/templates/python_samples/README.md [hunk 14]
```diff
@@ -0,0 +1,59 @@
+[//]: # "This README.md file is auto-generated, all changes to this file will be lost."
+[//]: # "To regenerate it, use `python -m synthtool`."
+
+## Python Sample Folder for {{ metadata['repo']['name_pretty'] }}
```
Comment (kolea2) [inline synthtool/gcp/templates/python_samples/README.md:4]: what do you think of simply `Python Samples for [product name]`?

File: synthtool/gcp/templates/python_samples/README.md [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+[//]: # "To regenerate it, use `python -m synthtool`."
+
+## Python Sample Folder for {{ metadata['repo']['name_pretty'] }} 
+
+This directory contains samples for {{ metadata['repo']['name_pretty'] }}, which may be used as a refererence for how to use this product. {% if metadata['repo']['custom_content'] is defined %}
+{{ metadata['repo']['custom_content']}}{% endif %}
+
+## Samples
+{% if metadata['repo']['samples']|length %}
+Samples, quickstarts, and other documentation available for this product is available at <a href="{{ metadata['repo']['product_documentation'] }}">the product documentation.</a>
```
Comment (kolea2) [inline synthtool/gcp/templates/python_samples/README.md:11]: Can we combine the first two sections? Something like (using bigtable as an example):
`This directory contains code samples for Cloud Bigtable. Samples, quickstarts, and other documentation available for this product can be found in the product documentation.`

**Review by kolea2 — COMMENTED** [review]
Thanks for adding the generated example, that was super helpful! Added a few additional comments based on that.

File: synthtool/gcp/templates/python_samples/README.md [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+</a>
+
+To run this sample:
+
+1. If this is your first time working with GCP products, you will need to set up [the Cloud SDK][cloud_sdk] or utilize [Google Cloud Shell][gcloud_shell]. This sample may [require authetication][authentication]{% if metadata['repo']['requires_billing'] %} and you will need to [enable billing][enable_billing]{% endif %}.
+
+1. Make a fork of this repo and clone the branch locally, then navigate to the sample directory you want to use.
+
+1. Install the dependencies needed to run the samples.
+        ```
```
Comment (billyjacobson) [inline synthtool/gcp/templates/python_samples/README.md:29]: Can we remove the $ for easier copy pasting as well as have these snippets on their own lines similar to how Node does it?  https://github.com/googleapis/nodejs-bigtable#installing-the-client-library

File: synthtool/gcp/templates/python_samples/README.md [hunk 17]
```diff
@@ -0,0 +1,59 @@
+[//]: # "This README.md file is auto-generated, all changes to this file will be lost."
+[//]: # "To regenerate it, use `python -m synthtool`."
+
+## Python Sample Folder for {{ metadata['repo']['name_pretty'] }} 
+
+This directory contains samples for {{ metadata['repo']['name_pretty'] }}, which may be used as a refererence for how to use this product. {% if metadata['repo']['custom_content'] is defined %}
```
Comment (billyjacobson) [inline synthtool/gcp/templates/python_samples/README.md:6]: There are a few places in the final result that say "this product" Could that be replaced by the product name?

**Review by billyjacobson — APPROVED** [review]
I think this looks good. We can always go an add more/clean up anything that needs updating after it's in the wild, but this is captures your original goals.

File: synthtool/gcp/templates/python_samples/README.md (hunk 14 above)
Comment (runargs) [reply synthtool/gcp/templates/python_samples/README.md:4]: Sounds good :)

File: synthtool/gcp/templates/python_samples/README.md (hunk 15 above)
Comment (runargs) [reply synthtool/gcp/templates/python_samples/README.md:11]: Updating!

File: synthtool/gcp/templates/python_samples/README.md (hunk 16 above)
Comment (runargs) [reply synthtool/gcp/templates/python_samples/README.md:29]: I thought I wasn't able to new lines of code blocks in but I checked up on how and fixed that up.

File: synthtool/gcp/common.py (hunk 12 above)
Comment (SurferJeffAtGoogle) [reply synthtool/gcp/common.py:65]: How does the code behave if you remove lines 63-65?

**Review by kolea2 — APPROVED** [review]

**Comment by kolea2** [comment]
LGTM. @SurferJeffAtGoogle do you have any additional comments?

File: synthtool/gcp/common.py (hunk 12 above)
Comment (runargs) [reply synthtool/gcp/common.py:65]: test_kwargs
I wrote a simple test to demonstrate what happens if there is null metadata. From what I can tell it seems like Jinja itself does not have the ability to handle undefined metadata.

Comment (runargs) [reply synthtool/gcp/common.py:65]: test_kwargs_code

**Review by SurferJeffAtGoogle — APPROVED** [review]

**Pull request merged**